core/src/main/scala/kafka/log/CleanerCache.scala

```diff
@@ -0,0 +1,316 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kafka.log
+
+import java.nio.ByteBuffer
+import java.security.MessageDigest
+import java.util
+
+import kafka.utils.{CoreUtils, nonthreadsafe}
+import org.apache.kafka.common.record.Record
+import org.apache.kafka.common.utils.{ByteUtils, Utils}
+
+trait CleanerCache {
+  /** The maximum number of entries this map can contain */
+  def slots: Int
+
+  /**
+   * Put this record in the cache, but only if it is greater than
+   * the currently associated record (if any).
+   * A record is considered to be greater than another if it has a larger version
+   * (or offset, if no version exists) than the currently cached record,
+   * or if no cached record exists at all.
+   *
+   * @param record  The record
+   * @return True if the record was inserted, false otherwise
+   */
+  def putIfGreater(record: Record): Boolean
+
+  /**
+   * Get the offset associated with this key.
+   *
+   * @param key The key
+   * @return The offset associated with this key or -1 if the key is not found
+   */
+  def offset(key: ByteBuffer): Long
+
+  /**
+   * Get the version associated with this key.
+   *
+   * @param key The key
+   * @return The version associated with this key or -1 if the key is not found
+   */
+  def version(key: ByteBuffer): Long
+
+  /**
+   * Sets the passed value as the latest offset.
+   * This value is not set if it is lesser than the currently kept value.
+   */
+  def updateLatestOffset(offset: Long)
+
+  /** Change the salt used for key hashing making all existing keys unfindable. */
+  def clear()
+
+  /** The number of entries put into the map (note that not all may remain). */
+  def size: Int
+
+  def utilization: Double = size.toDouble / slots
+
+  /** The latest offset put into the map. */
+  def latestOffset: Long
+
+  /**
+   * @return True if the passed record has a larger version (or offset, if no version exists)
+   *         than the currently cached record, or if no cached record exists at all
+   */
+  def greater(record: Record): Boolean
+}
+
+object Constants {
+  val OffsetStrategy: String = Defaults.CompactionStrategy
+  val TimestampStrategy: String = "timestamp"
+}
+
+/**
+ * A hash table used for deduplicating the log.
+ * This hash table uses a cryptographically secure hash of the key as a proxy
+ * for the key for comparisons and to save space on object overhead.
+ * Collisions are resolved by probing.
+ * This hash table does not support deletes.
+ *
+ * @param memory        The amount of memory this map can use
+ * @param hashAlgorithm The hash algorithm instance to use: MD2, MD5, SHA-1, SHA-256, SHA-384, SHA-512
```

Comment:
I don't know the Kafka code base, so please take this in context, but my feedback here is that using cryptographic hashes for dictionary hashing is not a good approach.  Its too slow.  Just use Java's native dictionary implementation.

Reply:
Hi @BertusGreeff . Please refer to the original code:
https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/log/OffsetMap.scala

You can, of course, file a change to address that.